kms/cloudkms/decrypter.go

```diff
@@ -0,0 +1,120 @@
+package cloudkms
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"errors"
+	"fmt"
+	"hash/crc32"
+	"io"
+
+	"cloud.google.com/go/kms/apiv1/kmspb"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+
+	"go.step.sm/crypto/kms/apiv1"
+	"go.step.sm/crypto/pemutil"
+)
+
+// CreateDecrypter implements the apiv1.Decrypter interface and returns
+// a crypto.Decrypter backed by a decryption key in Google Cloud KMS.
+func (k *CloudKMS) CreateDecrypter(req *apiv1.CreateDecrypterRequest) (crypto.Decrypter, error) {
+	if req.DecryptionKey == "" {
+		return nil, errors.New("decryption key cannot be empty")
+	}
+	return NewDecrypter(k.client, req.DecryptionKey)
+}
+
+// Decrypter implements a crypto.Decrypter using Google Cloud KMS.
+type Decrypter struct {
+	client        KeyManagementClient
+	decryptionKey string
+	publicKey     crypto.PublicKey
+}
+
+// NewDecrypter creates a new crypto.Decrypter backed by the given
+// Google Cloud KMS decryption key.
+func NewDecrypter(client KeyManagementClient, decryptionKey string) (*Decrypter, error) {
+	// Make sure that the key exists.
+	decrypter := &Decrypter{
+		client:        client,
+		decryptionKey: decryptionKey,
+	}
+	if err := decrypter.preloadKey(decryptionKey); err != nil { // TODO(hs): (option for) lazy load instead?
+		return nil, err
+	}
+
+	return decrypter, nil
+}
+
+func (d *Decrypter) preloadKey(signingKey string) error {
+	ctx, cancel := defaultContext()
+	defer cancel()
+
+	response, err := d.client.GetPublicKey(ctx, &kmspb.GetPublicKeyRequest{
+		Name: signingKey,
+	})
+	if err != nil {
+		return fmt.Errorf("cloudKMS GetPublicKey failed: %w", err)
+	}
+
+	d.publicKey, err = pemutil.ParseKey([]byte(response.Pem))
+	return err
+}
+
+// Public returns the public key of this decrypter
+func (d *Decrypter) Public() crypto.PublicKey {
+	return d.publicKey
+}
+
+// Decrypt decrypts ciphertext using the decryption key backed by Google Cloud KMS and returns
+// the plaintext bytes. An error is returned when decryption fails. Google Cloud KMS only supports
+// RSA keys with 2048, 3072 or 4096 bits and will always use OAEP. It supports SHA1, SHA256 and
+// SHA512. Labels are not supported.
+//
+// Also see https://cloud.google.com/kms/docs/algorithms#asymmetric_encryption_algorithms.
+func (d *Decrypter) Decrypt(rand io.Reader, ciphertext []byte, opts crypto.DecrypterOpts) ([]byte, error) {
+	if ropts, ok := opts.(*rsa.OAEPOptions); ok && ropts != nil {
+		if len(ropts.Label) > 0 {
+			return nil, errors.New("cloudKMS does not support RSA-OAEP label")
+		}
+		switch ropts.Hash {
+		case crypto.SHA1, crypto.SHA256, crypto.SHA512:
+			break
+		default:
+			return nil, fmt.Errorf("cloudKMS does not support hash algorithm %q with RSA-OAEP", ropts.Hash)
+		}
+	}
+	if _, ok := opts.(*rsa.PKCS1v15DecryptOptions); ok {
+		return nil, errors.New("cloudKMS does not support PKCS #1 v1.5 decryption")
+	}
```

Comment:
You can do something like the following:

```go
func validateOAEPOptions(o *rsa.OAEPOptions) error {
	if len(o.Label) > 0 {
		return errors.New("cloudKMS does not support RSA-OAEP label")
	}
	switch o.Hash {
	case crypto.Hash(0), crypto.SHA1, crypto.SHA256, crypto.SHA512:
		return nil
	default:
		return fmt.Errorf("cloudKMS does not support hash algorithm %q with RSA-OAEP", ropts.Hash)
	}
}

func (d *Decrypter) Decrypt(rand io.Reader, ciphertext []byte, opts crypto.DecrypterOpts) ([]byte, error) {
	if opts == nil {
		opts = &rsa.OAEPOptions{}
	}

	switch o := opts.(type) {
	case *rsa.OAEPOptions:
		if err := validateOAEPOptions(o); err != nil {
			return nil, err
		}
	case *rsa.PKCS1v15DecryptOptions:
		return nil, errors.New("cloudKMS does not support PKCS #1 v1.5 decryption")
	default:
		return nil, errors.New("invalid options for Decrypt")
	}
	// ...
}
```

You can also do nothing in the `case *rsa.OAEPOptions` and do the validation after the switch.

The algorithm really depends on the key itself, so it makes sense to allow crypto.Hash(0) too.

Reply:
Changed it in c72666284c34ac267f1a5a83d0dcebbfece6354b. Note the additional `o == nil` case. It's required in a fairly specific usage pattern.